In [37]:
import numpy as np
import pandas as pd

In [38]:
np.__version__

'1.23.3'

<h1> Question 1 </h1>
What's the version of NumPy that you installed?

<h3> 1.23.3 </h3>

In [39]:
df = pd.read_csv('data.csv')

In [40]:
df.shape

(11914, 16)

<h1> Question 2 </h1>

How many records are in the dataset?

<h3> 11914 </h3>

In [41]:
df.head()

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [150]:
df.Make.value_counts().head(3)

Chevrolet     1123
Ford           881
Volkswagen     809
Name: Make, dtype: int64

<h1> Question 3 </h1>

Who are the most frequent car manufacturers (top-3) according to the dataset?

<h3> Chevrolet, Ford, Volkswagen </h3>

In [43]:
new_df = df[df['Make'] == 'Audi']
new_df.head()

Make Model  Year  Engine Fuel Type  Engine HP  Engine Cylinders  \
17  Audi   100  1992  regular unleaded      172.0               6.0   
18  Audi   100  1992  regular unleaded      172.0               6.0   
19  Audi   100  1992  regular unleaded      172.0               6.0   
20  Audi   100  1992  regular unleaded      172.0               6.0   
21  Audi   100  1992  regular unleaded      172.0               6.0   

   Transmission Type      Driven_Wheels  Number of Doors Market Category  \
17            MANUAL  front wheel drive              4.0          Luxury   
18            MANUAL  front wheel drive              4.0          Luxury   
19         AUTOMATIC    all wheel drive              4.0          Luxury   
20            MANUAL  front wheel drive              4.0          Luxury   
21            MANUAL    all wheel drive              4.0          Luxury   

   Vehicle Size Vehicle Style  highway MPG  city mpg  Popularity  MSRP  
17      Midsize         Sedan           24        17        3105  2000  
18      Midsize         Sedan           24        17        3105  2000  
19      Midsize         Wagon           20        16        3105  2000  
20      Midsize         Sedan           24        17        3105  2000  
21      Midsize         Sedan           21        16        3105  2000

In [44]:
new_df.Model.nunique()

34

<h1> Question 4 </h1>

What's the number of unique Audi car models in the dataset?

<h3> 34 </h3>

In [45]:
df.isna().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

<h1> Question 5 </h1>

How many columns in the dataset have missing values?

<h3> 5 </h3>

In [151]:
engine_cyl_median_before_feeling = df['Engine Cylinders'].median()
engine_cyl_mode = df['Engine Cylinders'].mode()[0]

print('Median before feeling:', engine_cyl_median_before_feeling )

Median before feeling: 6.0


In [64]:
df.fillna({'Engine Cylinders': engine_cyl_mode},inplace=True)

In [68]:
engine_cyl_median_after_feeling = df['Engine Cylinders'].median()

print('Median after feeling:', engine_cyl_median_after_feeling )

Median after feeling: 6.0


<h1> Question 6 </h1>

Find the median value of "Engine Cylinders" column in the dataset. <br><br>
Next, calculate the most frequent value of the same "Engine Cylinders". <br><br>
Use the fillna method to fill the missing values in "Engine Cylinders" with the most frequent value from the previous step. <br><br>
Now, calculate the median value of "Engine Cylinders" once again. <br><br>
 
Has it changed?

<h3> NO!! </h3>

In [81]:
new_df_2 = df[
    df['Make']=='Lotus'
    ]
new_df_2 = new_df_2 [ ['Engine HP' , 'Engine Cylinders'] ]
new_df_2.shape

(29, 2)

In [82]:
new_df_2 = new_df_2.drop_duplicates()
new_df_2.shape

(9, 2)

In [88]:
x = new_df_2.to_numpy()
x.shape

(9, 2)

In [89]:
def dot_product(u,v):
    if u.shape != v.shape:
        return "They are not of the same shape!!!"

    result = 0
    size = u.shape[0]
    if len(u.shape) == 1:
        for i in range(size):
            result += u[i] * v[i]
    else:
        for i in range(size):
            result += dot_product(u[i],v[i])

    return result

In [146]:
def cross_product (u,v):
    num_row_u = u.shape[0]
    num_row_v = v.shape[0]
    try:
        num_col_u = u.shape[1]
    except:
        num_col_u = 1
    try:
        num_col_v = v.shape[1]
    except:
        num_col_v = 1

    result = np.zeros((num_row_u,num_col_v))

    if num_col_u == num_row_v:

        if num_col_v == 1:
            result = np.zeros(num_row_u)
            for row in range(num_row_u):
                result[row] = dot_product(u[row],v)
        else:
            for row in range(num_row_u):
                for col in range(num_col_v):
                    result[row][col] = dot_product(u[row],v[:,col])
        
        return result
    
    else:
        return print("The vectors can't be multiplied because: ", num_col_u , ' isnot equal to' , num_row_v)

In [99]:
np.set_printoptions(suppress=True)
result_matrix_matrix_mult = np.round(cross_product(x.T,x))
result_matrix_matrix_mult

array([[731684.,  13410.],
       [ 13410.,    252.]])

In [100]:
inverse = np.linalg.inv(result_matrix_matrix_mult)
inverse

array([[ 0.   , -0.003],
       [-0.003,  0.161]])

In [103]:
y = np.array([1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800])

In [149]:
m = cross_product(inverse,x.T)
w = cross_product(m,y)
w

array([  4.595, -63.564])

<h1> Question 7 </h1>

Select all the "Lotus" cars from the dataset.

Select only columns "Engine HP", "Engine Cylinders".

Now drop all duplicated rows using drop_duplicates method (you should get a dataframe with 9 rows).

Get the underlying NumPy array. Let's call it X.

Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

Compute the inverse of XTX.

Create an array y with values [1100, 800, 750, 850, 1300, 1000, 1000, 1300, 800].

Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

What's the value of the first element of w?

<h3> 4.5949 </h3>